Userendpoints

thquad opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
The feature lets users create endpoints if they have the "stratos.endpointadmin" scope.

## Description
<!--- Describe your changes in detail -->
### General Behaviour
- Users need the "stratos.endpointadmin" scope to use this feature. From now on, when i say users, assume they have the "stratos.endpointadmin" scope unless otherwise stated.
- In the config USER_ENDPOINTS_ENABLED can be set to various 
  - disabled: Disabled
  - admin_only: Admins will be able to see all endpoints, but users won't be able to see or create user endpoints.
  - enabled: Admins see all endpoints, users see all admin-endpoints and can create/edit/connect/see their own endpoints.
- Users can register endpoints, unless
  - An admin-endpoint with the same name / url already exists
  - A user-endpoint from the same user with the same name / url already exists  
- Users will have access to stratos, when no endpoints have been registered yet (the no-endpoint screen will be skipped)
- Users can edit, unregister and connect to their own user-endpoints
- User can't see user-endpoints created by other users
- ~~User-endpoints have the option for user-invites disabled. They can't be shared by users or admins~~
- Admins can edit and unregister user-endpoints, but won't be able to connect to them
- Admins can't create endpoints, if an endpoint with the same url / name already exists, unless
  - A checkbox has been added to the register screen to overwrite existing user-endpoints. When checked, user-endpoints with the same url will be unregistered

### Back-End
- Added file 20210201110000_Creator.go
  - Adds a new row to the cnsis table with name "creator"
    - User registers an endpoint -> value: user-id
    - Admin registers an endpoint -> value: ""
- config
  - Added an enum "USER_ENDPOINTS_ENABLED"
- middleware.go
  - Added `endpointMiddleware()`
    - Checks if user is admin or endpointadmin
    - if its not a CREATE request, then it also checks the existing endpoint and user role according to how it's descriped in section *General Behaviour*
- cnsi.go
  - `buildCNSIList()`
    - Checks if user is admin
      - Admin -> return all endpoints
      - User -> return own endpoints and admin-endpoints
      - User & enum set to admin_only -> return admin-endpoints
      - Enum set to disabled -> return admin-endpoints
  - `registerEndpoints()`
    - looks up user-id from session user and passes it on to DoRegisterEndpoints()
  - `DoRegisterEndpoints()`
    - Now accepts a user-id as parameter too. If parameter is "", it assumes an anonymous user.
    - Also checks now for described behaviours in section **General Behaviour**
    - For user-endpoints, the guid will now be generated with url+userID as input, instead of just the url
  - `unregisterCluster()`
    - Split into two functions: `unregisterCluster(c echo.Context)` and `doUnregisterCluster(cnsiGUID string)`, so that cnsi can be unregistered without the need of an `echo.Context`
  - Added `ListAdminEndpoints(userID string)`
    - returns all endpoints created by admins and given user-id
  - Added `listCNSIByAPIEndpoint(apiEndpoint string)`
    - returns all endpoints with the given APIEndpoint URL
  - Renamed `GetCNSIRecordByEndpoint()` and `cnsiRecordExists()` to `GetAdminCNSIRecordByEndpoint()` and `adminCNSIRecordExists()` to better reflect their function now
- authcnsi.go
  - `DoLoginToCNSI()`
    - Checks for described behaviour in section **General Behaviour**
- info.go
  - `getInfo()`
    - looks up saved id in column `creator` from cnsi and replaces it with interface.CreatorInfo
    - If no user can be found, the username will be set to just "user"
- main.go
  - set `ENABLE_USER_ENDPOINTS` to `disabled` as default
  - use different middleware if flag is enabled or not
- ~~userinvite/admin.go~~
  - ~~added a check for creator~~
- ~~userinvite/auth.go~~ 
  - ~~Added `checkEndpointCreator()`~~
    - ~~Checks if admin-endpoint, otherwise don't allow invites~~
- plugins/desktop/*
  - Adjust several files to also include new functions added to store structs
- structs.go
  - Add `Creator` to several structs
  - Add `UserEndpointsEnabled` to `PortalConfig` and `Info`
  - Add `OverwriteEndpoints` to `RegisterEndpointParams`
- pgsql_cnsis.go
  - Added `listCNSIsByCreator` sql query and belonging functions
  - Added function to query a list using `findCNSIByAPIEndpoint`
  - Created a generic `listBy()` function because DRY
- ~~pgsql_token.go~~
  - ~~Added `listAuthToken` sql query and belonging functions~~
- Tests
  - Adjusted mock data for several tests so they don't break

### Front-End
- endpoint.types.ts
  - added a `CreatorInfo` interface to hold creator user data for cnsi
- user-permission.directive.ts
  - Change so that it can have multiple permissions to check for, instead of just one. Results true if one permission return true.
- endpoint-list.helpers.ts
  - Show menu items according to described rules in section **General Behaviour**
- base-endpoints-data-source.ts
  - add default values for added parameters
- stratos-user-permissions.checker.ts
  - Add the `stratos.endpointadmin` as a scope that can be checked for
- session.service.ts
  - Add methods to return values of `USER_ENDPOINTS_ENABLED`
- auth.types.ts
  - add `enableUserEndpoints` to `interface SessionDataConfig`
- endpoint.actions.ts / endpoint.effects.ts / stratos-action-builders.ts
  - Add "overwriteEndpoints" variable 
- endpoints.service.ts
  - Let endpointadmins see stratos gui when no endpoints are there yet
- user.service.ts
  - Also save if user is an endpointadmin
- user-invite.service.ts
  - check if user-endpoint and disable invites
- endpoints-page.component
  - Show the add endpoint for admins and endpointadmins
- endpoint-card.component
  - Show who created the endpoint.
- add create-endpoint-helper component
  - extended by endpoint creation components
  - stores information about current endpoints in Stratos and if userendpoints are enabled
- create-endpoint-cf-step-1.component / git-registration.component
  - Add "Overwrite User Endpoints" checkbox to gui and form
- add table-cell-endpoint-creator component (and changed endpoint-list-config.service.ts accordingly)
  - Show the creator in the table view

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Issue https://github.com/cloudfoundry/stratos/issues/4753

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
OS: macOS Big Sur

Adjusted all unit tests with new model of endpoints (mock endpoints) so they don't fail.
Added back-end unit tests with this feature enabled for middleware, registering endpoints, connecting to endpoints and pgsql changes. 
Manually tested:
- Registering user-endpoint as user-A
- Editing user-endpoint as user-A
- Unregister user-endpoint as user-A
- Register same user-endpoint as user-A as user-B
- Overwrite user-endpoints as admin with checkbox "Automatically overwrite user endpoints"
- Unregister user-endpoint as admin
- Edit user-endpoint as admin
- Connect to admin-endpoint as user-A
- Connect to user-endpoint as user-A

Local minikube Kubernetes endpoints where used. 
Docker image splatform/stratos-uaa used as UAA.
Added users and necessary scopes with uaac.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Docs update
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have followed the guidelines in CONTRIBUTING.md, including the required formatting of the commit message

richard-cox reviewed (commented):
I've quickly run through the code and added some comments, haven't started testing though.

One question I had, which effects a couple of the comments, what was the reasoning for `User-endpoints have the option for user-invites disabled. They can't be shared by users or admins`?

I'm also actively started looking at the e2e failures, I believe these relate to the travis config or environment (they start to run fine locally)

thquad commented on src/frontend/packages/core/src/shared/components/list/list-types/endpoint/endpoint-list.helpers.ts:
```diff
@@ -107,10 +110,10 @@ export class EndpointListHelper {
-          row$
-        ).pipe(
+        createVisible: (row$: Observable<EndpointModel>) => combineLatest([
+            this.currentUserPermissionsService.can(StratosCurrentUserPermissions.EDIT_ENDPOINT),
```
This was an oversight from me, i will change it to the correct permission. Thank you.

thquad commented on src/frontend/packages/core/src/shared/components/list/list-types/endpoint/endpoint-list.helpers.ts:
```diff
@@ -132,11 +135,23 @@ export class EndpointListHelper {
+            row$
+          ]).pipe(
+            map(([userEndpointsEnabled, isAdmin, row]) => {
+              if (userEndpointsEnabled && !row.creator.admin && isAdmin){
```
This is hiding the connect button in user-endpoints for admins, so that admins can't connect to user created endpoints. Since users can't see endpoints created by other users, they are allowed to create an endpoint with the same APIUrl from another endpoint. Admins on the other side can see all user-endpoints. If I'm allowing admins to connect to user-endpoints, that would open up the possibility for admins to connect to the same APIUrl two or more times through different user-endpoints.

thquad commented on src/frontend/packages/core/src/shared/components/list/list-types/endpoint/endpoint-list.helpers.ts:
```diff
@@ -132,11 +135,23 @@ export class EndpointListHelper {
+            row$
+          ]).pipe(
+            map(([userEndpointsEnabled, isAdmin, row]) => {
+              if (userEndpointsEnabled && !row.creator.admin && isAdmin){
```
Yes, the creator is added in row 104 of info.go and should always return a creator. For legacy (and admin) endpoints, an anonymous admin creator will be returned.

thquad commented on src/jetstream/repository/interfaces/tokens.go:
```diff
@@ -10,6 +10,7 @@ type Token struct {
 // TokenRepository is an application of the repository pattern for storing tokens
 type TokenRepository interface {
 	FindAuthToken(userGUID string, encryptionKey []byte) (TokenRecord, error)
+	ListAuthToken(encryptionKey []byte) ([]TokenRecord, error)
```
This was used by a previous implementation, but i forgot to also remove the function in the TokenRepository. Since it's not used anymore, i will remove it.

thquad commented on src/frontend/packages/cloud-foundry/src/features/cf/user-invites/user-invite.service.ts:
```diff
@@ -143,13 +143,15 @@ export class UserInviteService {
+      map(([cf, auth, endpoint]) =>
         cf.global.isAdmin &&
-        auth.sessionData['plugin-config'] && auth.sessionData['plugin-config'].userInvitationsEnabled === 'true')
+        auth.sessionData['plugin-config'] && auth.sessionData['plugin-config'].userInvitationsEnabled === 'true' &&
```
"When registering an endpoint we show the option to 'share' it with other users. user specific endpoints should not be shared. this should be hidden if user is not an admin"

I changed it because of the first two sentences in the quote above (from your first reply to me in the issue). I assumed that the third sentence was still aimed at stratos roles and not cf roles. I will change it back.

thquad commented on src/jetstream/plugins/desktop/kubernetes/tokens.go:
```diff
@@ -14,6 +14,10 @@ func (d *TokenStore) FindAuthToken(userGUID string, encryptionKey []byte) (inter
 }
 
+func (d *TokenStore) ListAuthToken(encryptionKey []byte) ([]interfaces.TokenRecord, error) {
+	return d.store.ListAuthToken(encryptionKey)
```
"This was used by a previous implementation, but i forgot to also remove the function in the TokenRepository. I will remove it since it's not used anymore."

thquad commented on src/jetstream/main.go:
```diff
@@ -1112,11 +1118,20 @@ func (p *portalProxy) registerRoutes(e *echo.Echo, needSetupMiddleware bool) {
 	// API endpoints with Swagger documentation and accessible with an API key that require admin permissions
 	stableAdminAPIGroup := stableAPIGroup
-	stableAdminAPIGroup.Use(p.adminMiddleware)
+	if p.GetConfig().UserEndpointsEnabled != config.UserEndpointsConfigEnum.Disabled {
```
While implementing this, i encountered a problem that reminded me of an issue with the e2e tests.

From slack:
>The issue revolved around the /api/v1/endpoints request and using it in a system with no endpoints. we only use the request in e2e tests, in the browser we fetch endpoints via a /info request.

When i created the new group, i couldn't create any endpoints because the middleware for checking EDIT/DELETE routes also applied to the CREATE route. The cause was this issue: https://github.com/labstack/echo/issues/1519

Contrary to the docs, the middleware was applied to the whole group, instead of just the routes after the `Echo#Use()` function. 

I don't think this had any effect on the e2e tests tho, since the middleware is only applied when `UserEndpointsEnabled` is set to something else than `disabled`, but i wanted to mention this anyways.

thquad replied:
I noticed that the e2e test fail, because they are receiving a html instead of json. My assumption is, that my `stableAdminAPIGroup.Group("/endpoints")` overwrites the path `stableAPIGroup.GET("/endpoints", p.listCNSIs)` and therefore the backend serves static files. I'm not sure i can solve this without changing paths or putting the GET route also in the group. I will look into it and give a report later if i found a solution.

thquad commented on src/frontend/packages/core/src/features/endpoints/create-endpoint/create-endpoint-cf-step-1/create-endpoint-cf-step-1.component.ts:
```diff
@@ -151,4 +172,12 @@ export class CreateEndpointCfStep1Component implements IStepperStep, AfterConten
     this.showAdvancedOptions = !this.showAdvancedOptions;
   }
+
+  toggleOverwriteEndpoints() {
```
I've split it out into a superclass with the exception of`toggleOverwriteEndpoints`, since I would also need to move the `FormGroup` and all connected logic too.

thquad commented on src/jetstream/plugins/desktop/endpoints.go:
```diff
@@ -41,6 +41,14 @@ func (d *DesktopEndpointStore) FindByAPIEndpoint(endpoint string, encryptionKey
 	return d.store.FindByAPIEndpoint(endpoint, encryptionKey)
 }
 
+func (d *DesktopEndpointStore) ListByAPIEndpoint(endpoint string, encryptionKey []byte) ([]*interfaces.CNSIRecord, error) {
```
I've updated the functions.

richard-cox reviewed (changes requested):
I've given everything a rough test (ad-hoc list below). There's one blocking bug and some (debatable) blocking improvements. ~~There's also a test that I still need to execute.~~

Bug (blocking) - When the feature is set to admin_only endpoint admins can incorrectly create and update (their own previously created) endpoints. This only effects calling the api directly and can be tested by creating an API key for the required user and calling directly with something like ... `curl -k -H 'Authentication: Bearer <api key>' https://<backend url/api/v1/endpoints/<endpoint guid>`

Improvement (blocking) - The 'creator' value in an endpoint card or row is currently visible if the endpoint feature is not disabled. When feature is admin_only it should only show for admins (as users and endpoint admins won't see anything different than 'admin' creator)

Improvement (not blocking) - The differentiation between admin and user created endpoints could be improved, though I don't it has to be done at this time. 

~~Still to Test (blocking) - MYSQL Upgrade. I'm seeing some issues locally that's blocking the db upgrade test. I don;t think it's related to the PR but would like to still test upgrade before merging.~~ Tested upgrade from master --> userendpoints branch (runs migration 20210201110000_Creator) with existing endpoints connected by admin and non-admin users... all works fine


Feature Disabled
- ✓ admin - register/edit/connect/unregister
- ✓ non-admin - connect/disconnect
- ✓ creating an endpoint - urls should be unique
- ✓ shared endpoint
- ✓ kube endpoint
- ✓ metric endpoint
- ✓ helm endpoint
- ✓ github endpoint
- ✓ cf push (cf auto registered and users connect on log in) 
  - e2e cf... had to set CF_API_URL. tried using 10.x address to check 
- ✓ no endpoints connected - admin - redirected to endpoints page 
- ✓ no endpoints connected - endpoint admin - tough luck screen
- ✓ no endpoints connected - user - tough luck screen
- ✓ endpoint table & card - created info not visible
- ✓ API Key - Endpoint Admins can't list/create endpoints
- ✓ API Key - Admins can list/create endpoints
- ✓ backup/restore
- ✓ Test mysql upgrade

Feature Enabled
- ✓ admin - register/edit/connect/unregister admin endpoints
- ✓ admin - can overwrite endpoint admin endpoints (check jetstream log)
- ✓ admin - can't connect but can edit/unregister user endpoints
- ✓ admin - can't create own endpoint
- ✓ endpoint admin - register/edit/connect/unregister
- ✓ non-admin - connect
- ✓ creating an endpoint - urls should be unique.. unless admin overrides
- Endpoint visibility. 
  - ✓ users can't see each others endpoints
  - ✓ users can see their own
  - ✓ users can see admins
  - ✓ admins can see all
- ✓ API Key - Admins can list/create endpoints
- ✓ API Key - Endpoint Admins can list/create endpoints
- ✓ API Key - users can't list/create endpoints 
- ✓ no endpoints connected - admin - redirected to endpoints page
- ✓ no endpoints connected - endpoint admin - redirected to endpoints page
- ✓ no endpoints connected - user - tough luck screen
- ✓ endpoint table & card - created info visible

Feature Admin Only
- ✓ admin - register/edit/connect/unregister
- ✓ non-admin - connect
- Endpoint visibility. 
  - ✓ user can't see other users
  - ✓ users can't see their own
  - ✓ users can see admins
  - ✓ admins can see other users
- ✓ endpoint table & card - created info visible
- ✓ no endpoints connected - admin - redirected to endpoints page
- ✓ no endpoints connected - endpoint admin - tough luck screen
- ✓ no endpoints connected - user - tough luck screen
- ✓ API Key - Admins can list/create endpoints
- ✗ API Key - Endpoint Admins can't list/create endpoints
    - Endpoint Admin could create/edit their own previously created endpoints

thquad commented:
>Bug (blocking) - When the feature is set to admin_only endpoint admins can incorrectly create and update (their own previously created) endpoints. This only effects calling the api directly and can be tested by creating an API key for the required user and calling directly with something like ... curl -k -H 'Authentication: Bearer <api key>' https://<backend url/api/v1/endpoints/<endpoint guid>

I've changed the permission check and tested the call with an API key. Users can only modify their endpoints now when UserEndpoints is set to enabled.

>Improvement (blocking) - The 'creator' value in an endpoint card or row is currently visible if the endpoint feature is not disabled. When feature is admin_only it should only show for admins (as users and endpoint admins won't see anything different than 'admin' creator)

Is fixed now.

>Improvement (not blocking) - The differentiation between admin and user created endpoints could be improved, though I don't it has to be done at this time.

I've given you a suggestion on slack.

nwmac commented:
I'm not keen on the approach taken here:

- admins can not register personal endpoints
- if an admin registers an endpoint that a user has registered, their endpoint and connection is lost
- the endpoint IDs for a given endpoint are different depending on who registered it

I'd prefer the model where a user can register a personal endpoint. When we return the endpoint list to the front end, if the same endpoint exists for the user and the system, the user one takes precedence and is returned and the system one is effectively hidden. This will avoid the issues described above and is a clearer user experience.

thquad commented:
@nwmac Thank you for the reply.

> admins can not register personal endpoints
> if an admin registers an endpoint that a user has registered, their endpoint and connection is lost

Admins can get an additional checkmark during creation, so they can choose to create a personal endpoint. I can also remove the logic that endpoints will be overwritten by an admin. That way, users-endpoints will remain and hide system endpoints if they exist, like you described. 

> the endpoint IDs for a given endpoint are different depending on who registered it

I asked before if i could change how the guid is generated for users and the response was positive
https://github.com/cloudfoundry/stratos/issues/4753#issuecomment-776610057

I'm asking for this request to be moved to a separate issue since it would require some larger adjustments to the code (UPDATE & DELETE functions, tokens adjustments, cnsis adjustments, etc.). The remaining budget is limited and our client would need this feature as soon as possible.

thquad commented:
@nwmac 

I've made the following changes:
- Admins can now create a personal / user endpoint
- Personal endpoints created by admins are not seen by the users
- When system and user endpoints exist, then the system endpoint will be hidden from the user
- When a user / admin has an active connection to an API URL and connects to the same URL through another endpoint, then the connection will be moved to the new endpoint. This will prevent multiple connections to the same URL through different endpoints. (old endpoint will be disconnected)

richard-cox reviewed (changes requested):
One request and three minor comments, otherwise looks good!

The endpoint list (cards endpoint-card.component.ts  and table endpoints-list-config.service.ts) 'Creator' value for 'system' needs a minor tweak
- Endpoint Admins / Admins - Currently the value shows as just `system`, which could be a bit confusing. I'd suggest replacing `system` with either `System Endpoint` or `Global Endpoint`. We could do something more involved for the card... but the table would still need a similar treatment, so I think at this stage just the text as suggested should change.
- Standard User - Don't show 'Creator' field at all (they will only ever see 'system' endpoints)

thquad commented on website/sidebars.js:
```diff
@@ -31,7 +31,8 @@ module.exports = {
           'endpoints/cf/cf',
-          'endpoints/cf/invite-user-guide'
+          'endpoints/cf/invite-user-guide',
+          'endpoints/cf/user-endpoints'
```
Moved it to advanced.

thquad commented on src/jetstream/cnsi.go:
```diff
@@ -135,6 +203,11 @@ func (p *portalProxy) DoRegisterEndpoint(cnsiName string, apiEndpoint string, sk
 	newCNSI.SSOAllowed = ssoAllowed
 	newCNSI.SubType = subType
 
+	// admins currently can't create user endpoints
```
Removed.

thquad commented:
I've implemented the changes. I have noticed a little bug where the correct options weren't shown (edit, unregister) for users, so I fixed that too. I have changed the creator to `System Endpoint` to conform with the terminology used everywhere else.
